gr-zeromq/lib/base_impl.cc

```diff
@@ -166,29 +180,49 @@ namespace gr {
         return false;
 
       /* Reset */
-      d_msg.rebuild();
+      while (!d_msgs.empty()) {
+        d_msgs.pop();
+      }
       d_tags.clear();
-      d_consumed_items = 0;
-      d_consumed_bytes = 0;
+      d_total_consumed_items = 0;
+      d_total_consumed_bytes = 0;
+
+      boost::shared_ptr<struct rx_message> msg(new struct rx_message);
 
-      /* Get the message */
-      d_socket->recv(&d_msg);
+      /* Get the first message */
+      d_socket->recv(&msg->zmq_msg);
 
       /* Parse header */
       if (d_pass_tags)
       {
         uint64_t rcv_offset;
 
         /* Parse header */
-        d_consumed_bytes = parse_tag_header(d_msg, rcv_offset, d_tags);
+        msg->consumed_bytes = parse_tag_header(msg->zmq_msg, rcv_offset, d_tags);
 
         /* Fixup the tags offset to be relative to the start of this message */
         for (unsigned int i=0; i<d_tags.size(); i++) {
           d_tags[i].offset -= rcv_offset;
         }
       }
 
-      /* We got one ! */
+      d_msgs.push(msg);
+
+      /* Get the remaining messages */
+      int64_t more = 0;
+      size_t more_len = sizeof(more);
+      d_socket->getsockopt(ZMQ_RCVMORE, &more, &more_len);
+      while (more) {
+        msg.reset(new struct rx_message);
+        d_socket->recv(&msg->zmq_msg);
+        d_msgs.push(msg);
+
+        more = 0;
+        more_len = sizeof(more);
```

Comment:
Re-initializing more_len is unnecessary.  See the example at the bottom of http://api.zeromq.org/2-0:zmq-recv